docs/proposals/20210126-nc-and-ncp.md

```diff
@@ -0,0 +1,304 @@
+	// Log errors, Set Certificate Condition, Return
+}
+// Set Certificate Condition 
+```
+
+#### ControlPlaneEndpoint Setting
+
+After the certificates for the cluster have been found or generated we then need to check if the `ControlPlaneEndpoint` is set, if not we need to return and wait for the endpoint to be available. _Setting the `ControlPlaneEndpoint` field on the `NestedControlPlane` CR should be done by the `NestedAPIServer` resource._
+
+#### KUBECONFIG Management
+
+After the controller has checked that the `ControlPlaneEndpoint` is set, it moves on to verifying the `KUBECONFIG` has been created. To do this the controller will lookup the Secret again, `[cluster name]-kubeconfig`, If the secret is NOT found the controller will begin to generate the `KUBECONFIG` using the Cluster API `kubeconfig` helper functions.
+
+```go
+createErr := kubeconfig.CreateSecretWithOwner(
+	ctx,
+	r.Client,
+	clusterName,
+	endpoint.String(),
+	controllerOwnerRef,
+)
+```
+
+If the `KUBECONFIG` secret did exist the controller will check it's ownership, if it's missing owner references the kubeconfig will be `adopted` and owned by this control plane. Following ownership checks it will verify that the `KUBECONFIG` certificates are still valid, following CAPI's design this will default to 1 year expiry and a 6 month rotation timing, if those need rotation that operation will be done and the secret is updated.
+
+```go
+needsRotation, err := kubeconfig.NeedsClientCertRotation(configSecret, certs.ClientCertificateRenewalDuration)
+if err != nil {
+	// Return errors
+}
+
+if needsRotation {
+	if err := kubeconfig.RegenerateSecret(ctx, r.Client, configSecret); err != nil {
+		// Return errors
+	}
+}
+```
+
+## Upgrade Strategy
+
+Both `NestedCluster` & `NestedControlPlane` resources aren't actually responsible for the upgrading of the physical components, this is left to the downstream component controllers. When they receive an updated CR for `NestedEtcd` or `NestedAPIserver` those controllers are responsible for keeping the control planes updated. The `NestedControlPlane` and `NestedCluster` both need to be kept in-sync if update do happen to make sure `spec.controlPlaneEndpoint` is still pointed at the right access.
+
+The exception to this rule is the certificate management, because the control planes get their certificates through the `NestedControlPlane` controller creating the artifacts and storing in Kubernetes secrets (if not supplied) then we need to make sure the secrets receive that same updated secret when the certificates expire and are renewed, this is an action the component controllers MUST watch for but COULD be triggered via NCP.
+
+
+### Version Skew Strategy 
+
+Version skew is a problem that can easily happen, since the NCP doesn't prescribe the overall version of the stack, these musy me managed manually by the individual component controllers and the 
```

Comment:
The sentence is chopped.

Reply:
oops, yeah forgot to finish that. thanks.

Reply:
Updated.